thriftpy2/contrib/aio/protocol/compact.py

```diff
@@ -0,0 +1,318 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import absolute_import
+
+import asyncio
+from struct import unpack
+
+from thriftpy2.protocol.exc import TProtocolException
+from thriftpy2.thrift import TException, TType
+from thriftpy2.protocol.compact import (
+    from_zig_zag,
+    CompactType,
+    TCompactProtocol,
+)
+
+
+@asyncio.coroutine
+def read_varint(trans):
+    result = 0
+    shift = 0
+
+    while True:
+        x = yield from trans.read(1)
+        byte = ord(x)
+        result |= (byte & 0x7f) << shift
+        if byte >> 7 == 0:
+            return result
+        shift += 7
+
+
+class TAsyncCompactProtocol(TCompactProtocol):  # Inherit all of the writing
+    """Compact implementation of the Thrift protocol driver."""
+    PROTOCOL_ID = 0x82
+    VERSION = 1
+    VERSION_MASK = 0x1f
+    TYPE_MASK = 0xe0
+    TYPE_BITS = 0x07
+    TYPE_SHIFT_AMOUNT = 5
+
+    @asyncio.coroutine
+    def _read_size(self):
+        result = yield from read_varint(self.trans)
+        if result < 0:
+            raise TException("Length < 0")
+        return result
+
+    @asyncio.coroutine
+    def read_message_begin(self):
+        proto_id = yield from self.read_ubyte()
+        if proto_id != self.PROTOCOL_ID:
+            raise TProtocolException(TProtocolException.BAD_VERSION,
+                                     'Bad protocol id in the message: %d'
+                                     % proto_id)
+
+        ver_type = yield from self.read_ubyte()
+        type = (ver_type >> self.TYPE_SHIFT_AMOUNT) & self.TYPE_BITS
+        version = ver_type & self.VERSION_MASK
+        if version != self.VERSION:
+            raise TProtocolException(TProtocolException.BAD_VERSION,
+                                     'Bad version: %d (expect %d)'
+                                     % (version, self.VERSION))
+        seqid = yield from read_varint(self.trans)
+        name = yield from self.read_string()
+        return name, type, seqid
+
+    @asyncio.coroutine
+    def read_message_end(self):
+        assert len(self._structs) == 0
+
+    @asyncio.coroutine
+    def read_field_begin(self):
+        type = yield from self.read_ubyte()
+        if type & 0x0f == TType.STOP:
+            return None, 0, 0
+
+        delta = type >> 4
+        if delta == 0:
+            fid = from_zig_zag((yield from read_varint(self.trans)))
+        else:
+            fid = self._last_fid + delta
+        self._last_fid = fid
+
+        type = type & 0x0f
+        if type == CompactType.TRUE:
+            self._bool_value = True
+        elif type == CompactType.FALSE:
+            self._bool_value = False
+
+        return None, self._get_ttype(type), fid
+
+    @asyncio.coroutine
+    def read_field_end(self):
+        pass
+
+    @asyncio.coroutine
+    def read_struct_begin(self):
+        self._structs.append(self._last_fid)
+        self._last_fid = 0
+
+    @asyncio.coroutine
+    def read_struct_end(self):
```

Comment:
Seems like there are some methods do not need to add `asyncio.coroutine` decorator.

Reply:
I decorated all the methods starting with read to make it easier to convert by assuming all reads were coroutines. I can change it if you like as there is a slight performance cost.

Reply:
Yes, if the method does not depend on I/O, it is better to remove the coroutine decorator.

Reply:
I removed all the ones I could. TAsyncClient assumes `read_message_end()` is a coroutine so I left that one with the decorator.

Reply:
It may be worth it to go through all the different protocol classes and mark any helper methods as private so it's easier to determine which need to be async regardless of content based upon the public interface.

Reply:
Yes, maybe we should promise an interface to describe which methods of a protocol should be coroutines.

Reply:
I can take care of that in a dedicated PR for standardizing interfaces.